src/Database/LSMTree/Internal/IndexOrdinaryAcc.hs

```diff
@@ -0,0 +1,100 @@
+                     (SerialisedKey (SerialisedKey'))
+import           Database.LSMTree.Internal.Vector (mkPrimVector)
+
+{-|
+    A general-purpose fence pointer index under incremental construction.
+
+    A value @IndexOrdinaryAcc lastKeysRevRef baler@ denotes a partially
+    constructed index with the following properties:
+
+      * The keys pointed to by @lastKeysRevRef@ are the keys that the index
+        assigns to pages, but in reverse order.
+      * The @baler@ object is used by the index for incremental output of the
+        serialised key list.
+-}
+data IndexOrdinaryAcc s = IndexOrdinaryAcc !(STRef s [SerialisedKey]) !(Baler s)
+
+-- | Creates a new, initially empty, index.
+new :: Int                       -- ^ Maximum chunk size in bytes
+    -> ST s (IndexOrdinaryAcc s) -- ^ Construction of the index
+new maxChunkSize = IndexOrdinaryAcc <$> newSTRef [] <*> createBaler maxChunkSize
+
+{-
+    Appends a single key to the key list of an index and outputs newly available
+    chunks of the serialised key list.
+-}
+appendKey :: SerialisedKey -> IndexOrdinaryAcc s -> ST s [Chunk]
+appendKey lastKey@(SerialisedKey' lastKeyBytes)
+          (IndexOrdinaryAcc lastKeysRevRef baler)
+    = do
+          modifySTRef lastKeysRevRef (lastKey :)
+          chunksIntoLastKeySize <- feedBaler lastKeySizeBytes baler
+          chunksIntoLastKey <- feedBaler lastKeyBytes baler
+          return (chunksIntoLastKeySize ++ chunksIntoLastKey)
+    where
+
+    lastKeySize :: Int
+    lastKeySize = Primitive.length lastKeyBytes
+
+    lastKeySizeAsWord16 :: Word16
+    lastKeySizeAsWord16
+        | lastKeySize <= fromIntegral (maxBound :: Word16)
+            = fromIntegral lastKeySize
+        | otherwise
+            = error "Serialised key too large"
+
+    lastKeySizeBytes :: Primitive.Vector Word8
+    lastKeySizeBytes = mkPrimVector 0 2 $
+                       byteArrayFromListN 1 [lastKeySizeAsWord16]
+
+{-|
+    Appends keys to the key list of an index and outputs newly available chunks
+    of the serialised key list.
+-}
+append :: Append -> IndexOrdinaryAcc s -> ST s [Chunk]
+append (AppendSinglePage _ lastKey)    index = appendKey lastKey index
+append (AppendMultiPage key pageCount) index = fmap concat                $
+                                               sequence                   $
+                                               genericReplicate pageCount $
+                                               appendKey key index
+
+{-|
+    End an incremental construction of an index.
+
+    This returns the constructed index, along with a final chunk in case the
+    serialised key list has not been fully output yet.
+-}
+end :: IndexOrdinaryAcc s -> ST s (Maybe Chunk, IndexOrdinary)
+end (IndexOrdinaryAcc lastKeysRevRef baler) = do
+    lastKeysRev <- readSTRef lastKeysRevRef
+    remnant <- readBalerRemnant baler
+    return (remnant, IndexOrdinary (fromList (reverse lastKeysRev)))
```

Comment:
`fromList (reverse lastKeysRev))` seems like a relatively costly operation. It might be better to use a mutable vector, as mentioned before.

Reply:
Note: we know the *maximum* number of keys beforehand. It's a parameter given to the `RunAcc`, and from there it could also be sent to this construction code. As such, we can pre-allocate the mutable vector

https://github.com/IntersectMBO/lsm-tree/blob/main/src/Database/LSMTree/Internal/RunAcc.hs#L85-L86

Reply:
How tight is this upper bound expected to be? Can we be reasonably sure that we won’t waste too much space with this approach? Is the bound maybe even so tight that we can just use the contents of that mutable vector for the finally constructed index without copying?

Reply:
This upper bound isn't super tight. During a merge, resolving `n` duplicate keys produces *at most* 1 key -- for intermediate levels, it's going to be exactly `1`, on the last level it can be `0` or `1`. 

We can avoid wasting too much space by doing a [`force`](https://hackage.haskell.org/package/vector-0.13.1.0/docs/Data-Vector-Primitive.html#v:force)

Reply:
Changed in f84d4de5af557746cf3b9471acc790f1dd6e7218 and 33ceb2d0bbb019f6543ebd982943bdc6a116588a.